### ISSUE TITLE
Incorrect behavior in test case with complex "with ... as ..." statements.

### ISSUE THREAD

**Issue by anadon** [open-policy-agent/opa#1173]
# Expected Behavior
When running `opa test` on the attached file, run the included test without error or provide a valid error message.

# Actual Behavior
When the test is executed, it errors out claiming the test references itself.

# Steps to Reproduce the Problem

  1. `opa test . -v`

# Additional Info

```
[anadon@goodadvicemallard src]$ ./../bin/opa test bug/ -v
1 error occurred: bug/package arroyo_policy_upload.rego:24: rego_recursion_error: rule test_one is recursive: test_one -> test_one
```

May be related to #799 

[arroyo_policy_upload.zip](https://github.com/open-policy-agent/opa/files/2765798/arroyo_policy_upload.zip)

**Comment by anadon** [comment]
My boss figured out a bit more.

Works:
```
package arroyo_policy

# import things

available_roles = ["USER", "ADMIN", "OWNER"]

account_status = ["ENABLED", "DISABLED"]


default allow = false

allow {
 input.method = "GET"
 input.path = ["orgs", org_id]
 data.user[i].id = input.user_id
 data.user[i].status = "ENABLED"
 data.user[i].orgs[j].id = org_id
 data.user[i].orgs[j].roles[_] = "ADMIN"
}


user = [{
               "id": "test_user",
               "orgs": [{
                       "id": "arroyo-test-org",
                       "roles": ["USER", "ADMIN"]
               }],
               "status": "ENABLED",
}]

test_one {
 allow with input as {"method": "GET", "path": ["orgs", "arroyo-test-org"], "user_id" : "test_user"}
       with data.user as user
}
```

Doesn't work:
```
package arroyo_policy

# import things

available_roles = ["USER", "ADMIN", "OWNER"]

account_status = ["ENABLED", "DISABLED"]


default allow = false

allow {
 input.method = "GET"
 input.path = ["orgs", org_id]
 data.user[i].id = input.user_id
 data.user[i].status = "ENABLED"
 data.user[i].orgs[j].id = org_id
 data.user[i].orgs[j].roles[_] = "ADMIN"
}



nope = {"user": [{
               "id": "test_user",
               "orgs": [{
                       "id": "arroyo-test-org",
                       "roles": ["USER", "ADMIN"]
               }],
               "status": "ENABLED",
}]}


test_one {
 allow with input as {"method": "GET", "path": ["orgs", "arroyo-test-org"], "user_id" : "test_user"}
       with data as nope
}
```


And works:
```
package arroyo_policy

# import things

available_roles = ["USER", "ADMIN", "OWNER"]

account_status = ["ENABLED", "DISABLED"]


default allow = false

allow {
 input.method = "GET"
 input.path = ["orgs", org_id]
 data.user[i].id = input.user_id
 data.user[i].status = "ENABLED"
 data.user[i].orgs[j].id = org_id
 data.user[i].orgs[j].roles[_] = "ADMIN"
}

nope = {"user": [{
               "id": "test_user",
               "orgs": [{
                       "id": "arroyo-test-org",
                       "roles": ["USER", "ADMIN"]
               }],
               "status": "ENABLED",
}]}


test_one {
 allow with input as {"method": "GET", "path": ["orgs", "arroyo-test-org"], "user_id" : "test_user"}
       with data.user as nope.user
}
```

**Comment by tsandall** [comment]
TLDR; this is working as expected. I'm going to close this for now.

The rule is considered recursive because it's referring to itself implicitly in `with data ...`. All policies are implicitly rooted under `data`. E.g., to ask for the value of `allow` or `test_one` you would write `data.arroyo_policy.allow` or `data.arroyo_policy.test_one`. Since `data` (or `data.arroyo_policy`) would cover `test_one`, you see this error.

**Comment by murphyke** [comment]
I think the docs should mention this, since it is a gotcha that may make a bad impression on the uninitiated.

**Comment by tsandall** [comment]
One thing that could help would be to include more context in the error message. For example, we currently throw away the information about where the recursion was introduced and report the error on the rule that contains the recursion. If the recursion error was more like type errors it might help...

```
1 error occurred: bug/package arroyo_policy_upload.rego:26: rego_recursion_error: rule test_one is recursive: test_one -> test_one
        with data as {"user":
             ^^^^
```
